feat: concrete material properties and material class for Eurocode 2, 2004

mortenengen opened the pull request:
# Description
Implementation of the material properties of concrete according to Eurocode 2, Tab. 3.1, and a `ConcreteEC2_2004` material class.

# To do
- [x] Implement material properties:
  - [x] fcm
  - [x] fctm
  - [x] fctk_5
  - [x] fctk_95
  - [x] Ecm
  - [x] eps_c1
  - [x] eps_cu1
  - [x] eps_c2
  - [x] eps_cu2
  - [x] n
  - [x] eps_c3
  - [x] eps_cu3
- [x] Implement material class

mortenengen commented:
Hi @DanielGMorenaFhecor, please have a look at this PR and let me know if you have any comments.

DanielGMorenaFhecor reviewed (changes requested):
Good job Morten! **I think I got your top-to-bottom workflow** in which you design the parent `ConcreteEC2_2004` and then move to the bottom to implement the specifics of the EC2 2004!

### I got inspired by the way you created this code
Just as a way of **making more consistent Concrete, Sections, Elements,… classes in the future and also save time in the development time (for us and newcomers)**. **A way of creating a more scoped and time-bounded development steps.**  
What do you all think about this workflow **(user friendly diagram below)**:

1. **We design the architecture** (that you Morten and Diego through the meetings have already done!)
2. When we are ready for a concrete implementation (in coding terms, not material) then, **we can assign that implementation to someone** and review alltogether if it is correct and meets all the requirements. Just as we do also sometimes in the meetings. As we did with the `ConcreteEC2_2024` class. We can assign one team member to review the rest of the implementations.
3. Now, we can create Issues in the repo for the all the other implementations `ConcreteEC2_2004`, `ConcreteAASHTO`,… and in the issue, relate to the `ConcreteEC2_2024` suggesting to mimic its structure, like ‘almost copying’ it but with the specifics and namings.
4. Because we are already basing our code in a ready-to-go class I think we manage to get **quicker and more consistent implementations** and an easier onboarding for newcomers.
5. This may also lead to other more efficient way of organizing the scope of the structural codes developments. **Instead of devour all the formulas in the codes** with all the documentation and tests, which **it takes a ton of time** (and LET'S BE HONEST, **NOT REALLY FUN**), we just develop the formulas that we use in the upper structure.

Here it is a more user friendly diagram:
![image](https://github.com/fib-international/structuralcodes/assets/80036012/7a1629c5-dfb8-4428-8d2d-575356546dc9)


### Now, the LESS general topics: related to your code

- [x] In some `ec2_2024`definitions I missed some units. I know that in most cases people will use MPa, but new users that will just start to use the package… well I do not know if they are going to go into the depths of the code to seek which units they need to use. I think we should specified the units or, when necessary, say `consistent units`.

- [ ] I would add the formula to compute the stress of concrete (parabolic and linear), that could also be nice to be an option for the `ConcreteEC2_2004` class that we can pass as a string in the constructor. So the user can pick the constitutive law.

- [x] The definition of `fcd` is missing in the `ConcreteEC2_2004` class. Here we come up again with some time dependent properties and the security factors that we already discussed in other classes.


### Thank you for your work!

mortenengen commented on structuralcodes/codes/ec2_2004/_concrete_material_properties.py:
```diff
@@ -0,0 +1,148 @@
+from structuralcodes.codes import mc2010
+
+
+def fcm(fck: float, delta_f: float = 8) -> float:
```
Good point. I suggest fixes in 99dcff4.

mortenengen commented on structuralcodes/codes/ec2_2004/_concrete_material_properties.py:
```diff
@@ -0,0 +1,148 @@
+    EN 1992-1-1:2004, Table 3.1.
+
+    Args:
+        fck (float): The caracteristic compressive strength of concrete.
```
Typo fixed in 99dcff4 😃

mortenengen commented on structuralcodes/codes/ec2_2004/_concrete_material_properties.py:
```diff
@@ -0,0 +1,148 @@
+    EN 1992-1-1: 2004, Table 3.1.
+
+    Args:
+        fck (float): The caracteristic compressive strength of concrete.
```
Typo fixed in 99dcff4 😃

mortenengen commented on structuralcodes/codes/ec2_2004/__init__.py:
```diff
@@ -2,6 +2,20 @@
+    fctk_5,
+    fctk_95,
+    fctm,
+    n,
```
I agree. I have suggested a change in 3f9b356.

mortenengen commented on structuralcodes/codes/ec2_2004/_concrete_material_properties.py:
```diff
@@ -0,0 +1,148 @@
+    Args:
+        _fctm (float): The mean tensile strength of concrete.
+    """
+    return mc2010.fctkmin(_fctm=_fctm)
```
Good point. To be consistent, I have forced absolute for all args in 1c56d8a.

mortenengen commented:
Thanks for the thorough review, @DanielGMorenaFhecor. I have addressed your direct comments above.

You are right that fcd was missing from the implementation, thanks for pointing it out. I have added a simple implementation of it in 82be739, including tests, and a method on `ConcreteEC2_2004`. The method on the concrete class is deliberately made simple, since we anyway need to figure out a clever way to handle the partial factors and other parameters that we get from the national annexes. We will return to this 😃 

I also deliberately did not implement a function for calculating stress according to the parabolic rectangular law. We have an implementation of this law in #27, and after merging with that branch, we should make sure that all our material classes have proper methods for handling constitutive laws. So, we will return to this as well 😃

mortenengen commented:
Also, thanks for your general comment related to our development workflow! I think what you have sketched makes perfect sense, since as you say, it might speed up the process. I do sense a bit of impatience, though, and this I can understand. We are moving slowly, but steadily. We are carefully considering our steps, but we are also bold in taking leaps. I am the first to agree that writing tests and docstrings is not a ton of fun. But one thing that is even less fun, is to break existing code without knowing it. One reason for the slow pace, is that we have prioritized building a good test suite. For the moment, we have approximately 1300 tests and a coverage of 96%! And thanks to you and others, we are on the way to build the sustainable library with the strong integrity we set out to build. I foresee that after merging #27, we can start moving slightly faster. Let us see how we can implement your ideas then 😃

DanielGMorenaFhecor reviewed (approved):
Great job!

As you said, we will go into the details of the `fcd` implementation when we handle the National Annexes! 👍
